boxcli/midcobra/telemetry.go

```diff
@@ -70,15 +74,19 @@ func (m *telemetryMiddleware) postRun(cmd *cobra.Command, args []string, runErr
 		return // Ignore invalid commands
 	}
 
-	trackEvent(segmentClient, &event{
+	evt := &event{
 		AppName:     m.opts.AppName,
 		AppVersion:  m.opts.AppVersion,
 		Command:     subcmd.CommandPath(),
 		CommandArgs: subargs,
 		DeviceID:    deviceID(),
 		Duration:    time.Since(m.startTime),
-		Failed:      runErr != nil,
-	})
+	}
+	if runErr != nil {
+		evt.Failed = true
+		evt.FailedMsg = runErr.Error()
```

Comment:
It might be a good idea to only track a distinct error type so that we can redact errors for analytics. Otherwise I worry that we might return an error with sensitive info somewhere else in the code and accidentally log it.

Maybe something like this?

```go
type RedactedError interface {
	Redact() string
}

func (m *telemetryMiddleware) postRun(cmd *cobra.Command, args []string, runErr error) {
	// ...

	if runErr != nil {
		evt.Failed = true
		if redactErr, ok := runErr.(RedactedError); ok {
			evt.FailedMsg = redactErr.Redact()
		}
	}
}
```

Reply:
I like the idea, but I'm also worried this means we will *not* capture a bunch of errors as a result. I'm trying to think if there's an in-between.

Reply:
Another option is to do redacted format strings.

```go
// Not sure on the naming or if it should go in a new package.
type SafeError struct {
	format string
	args   []any
}

func (e *SafeError) Redacted() string {
	// This is ugly. You could probably do something fancier by substituting
	// the args with something else of the same type. Or use a template string
	// instead of fmt.
	return format
}

func (e *SafeError) Error() string {
	return fmt.Sprintf(e.format, e.args...)
}
```